### ISSUE TITLE
Transactions with a create and an update to the same document causes a `version_conflict_engine_exception`

### ISSUE THREAD

**Issue by yoitsro** [rwynn/monstache#552]
Originally related to #514, but different because this is when change streams are being used.

My monstache setup is in HA mode with two workers. I have change streams enabled on specific collections.

So within my application the following happens:
- Create a transaction
- Create a new document
- Update that document
- Commit the transaction

What I can see within the monstache worker which picked this up (with each log as a separate code block and the JSON formatted):

<details>
<summary>Create request</summary>

```
TRACE 2021/08/18 20:40:09 POST /_bulk HTTP/1.1
Host: elasticsearch:9200
User-Agent: elastic/7.0.22 (linux-amd64)
Content-Length: 631
Accept: application/json
Authorization: Basic ZWxhc3RpYzpFc1Bhc3N3b3JkMTIzNCE=
Content-Type: application/x-ndjson
Accept-Encoding: gzip

{
  "index": {
    "_index": "index1",
    "_id": "5f686b901aba0200136f45b1",
    "version": 6997872717399588865,
    "version_type": "external"
  }
}
{
  "__v": 0,
  "created_at": "2020-09-21T09:00:00.000Z",
  "ends_at": "2020-09-21T16:00:00.000Z",
  "members": [],
  "starts_at": "2020-09-21T08:00:00.000Z",
  "status": 1,
  "updated_at": "2020-09-21T09:00:00.000Z"
}
```

</details>

<details>
<summary>Update request</summary>

```
TRACE 2021/08/18 20:40:09 POST /_bulk HTTP/1.1
Host: elasticsearch:9200
User-Agent: elastic/7.0.22 (linux-amd64)
Content-Length: 962
Accept: application/json
Authorization: Basic ZWxhc3RpYzpFc1Bhc3N3b3JkMTIzNCE=
Content-Type: application/x-ndjson
Accept-Encoding: gzip

{
  "index": {
    "_index": "index1",
    "_id": "5f686b901aba0200136f45b1",
    "version": 6997872717399588865,
    "version_type": "external"
  }
}
{
  "__v": 1,
  "created_at": "2020-09-21T09:00:00.000Z",
  "ends_at": "2020-09-21T16:00:00.000Z",
  "members": [
    {
      "created_at": "2020-09-21T09:00:00.000Z",
      "ends_at": "2020-09-21T16:00:00.000Z",
      "id": "91effce3-3e29-4036-80bc-b94f8278734e",
      "starts_at": "2020-09-21T09:00:00.000Z",
      "user": "5f686b901aba0200136f45ac"
    }
  ],
  "starts_at": "2020-09-21T08:00:00.000Z",
  "status": 1,
  "updated_at": "2020-09-21T09:00:00.000Z"
}
```

</details>

<details>
<summary>Create response</summary>

```
TRACE 2021/08/18 20:40:09 HTTP/1.1 200 OK
Content-Length: 268
Content-Type: application/json; charset=UTF-8

{
  "took": 8,
  "errors": false,
  "items": [
    {
      "index": {
        "_index": "index1",
        "_type": "_doc",
        "_id": "5f686b901aba0200136f45b1",
        "_version": 6997872717399588865,
        "result": "created",
        "_shards": {
          "total": 1,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 2,
        "_primary_term": 1,
        "status": 201
      }
    }
  ]
}
```
</details>

<details>
<summary>Update response</summary>

```
TRACE 2021/08/18 20:40:09 HTTP/1.1 200 OK
Content-Length: 429
Content-Type: application/json; charset=UTF-8

{
  "took": 3,
  "errors": true,
  "items": [
    {
      "index": {
        "_index": "index1",
        "_type": "_doc",
        "_id": "5f686b901aba0200136f45b1",
        "status": 409,
        "error": {
          "type": "version_conflict_engine_exception",
          "reason": "[5f686b901aba0200136f45b1]: version conflict, current version [6997872717399588865] is higher or equal to the one provided [6997872717399588865]",
          "index_uuid": "mjn52OJhS5y9wK80pn-XEw",
          "shard": "0",
          "index": "index1"
        }
      }
    }
  ]
}
```
</details>

Based on this output, it would appear that the create and update operations both have the same version, which makes sense because the version is based off of the mongo timestamp (I think), which is the same between the create and update because they're both committed at the same time as part of the transaction.

What do you think can be done here?

One thing I can think of is providing monstache with a way of "deduplicating" changes with the same version. In my case, it would be to take the document with the highest `__v` value, but I understand this would have a performance penalty. Then again, consistent, accurate replication from mongo to elasticsearch is probably much more important than speed. After all, missing data is the worst case scenario.

**Comment by yoitsro** [comment]
Just found a potentially better way of deduplicating changes. According to the [mongo docs](https://docs.mongodb.com/manual/changeStreams/#event-notification), it looks like events associated with a transaction come with a `txnNumber`. That should give an order of the events which affect a single document.

**Comment by rwynn** [comment]
Hi @yoitsro yeah this is tricky.  I think it might be possible to work around this by adding the `txnNumber` from the change event to the version number.  What do these values look like for the example you have?  Are these unique for each change event in the txn and increasing?

**Comment by rwynn** [comment]
Monstache does rely on the version number to ensure updates to Elasticsearch are done correctly when multiple events happen nearly simultaneously on a document.

**Comment by rwynn** [comment]
Based on this answer I'm not sure `txnNumber` would help us here:

https://stackoverflow.com/questions/65038952/get-all-documents-involved-in-a-transaction-when-using-change-streams-in-mongodb

Because it would be the same number for Create document / Update document since they are part of the same transaction.

**Comment by yoitsro** [comment]
Yeah, I came to this conclusion after a bit of digging too. Sorry for not letting you know!

We can't necessarily rely on the timestamp it was received at either... Looking at [all the data](https://docs.mongodb.com/manual/reference/change-events/#std-label-change-stream-output) returned as part of a change stream, it doesn't look like there's any way to infer the order of the operation as part of the transaction...

Any other thoughts @rwynn?

**Comment by rwynn** [comment]
hey @yoitsro the only other thought I have is to allow configuring Monstache to add 1 to the version number (derived from the clusterTime of the change event) of all update operations.  Would probably also might want to add 1 to delete options in this case also since a document could potentially be deleted in the same txn.  Do you think that would work?

**Comment by rwynn** [comment]
I suppose to revise I would go with

insert version = timestamp
update version = timestamp + 1
delete version = timestamp + 2

that should handle most txn scenarios right?

c -> u -> d
c -> d
c -> u

**Comment by yoitsro** [comment]
That sounds like a clever workaround.

There are maybe some theoretical issues which could occur but nothing which I can imagine would surface in practice.
